File: java-project-euler-100/src/main/java/net/talaatharb/projecteuler100/P017.java
package net.talaatharb.projecteuler100;

import java.util.HashMap;
import java.util.Map;

/**
 * P017
 * 
 * @author mharb
 *
 */
public class P017 extends Solution {

	private final Map<Integer, String> memoryCache;

	/**
	 * Public constructor
	 */
	public P017() {
		memoryCache = new HashMap<>();
		memoryCache.put(1, "one");
		memoryCache.put(2, "two");
		memoryCache.put(3, "three");
		memoryCache.put(4, "four");
		memoryCache.put(5, "five");
		memoryCache.put(6, "six");
		memoryCache.put(7, "seven");
		memoryCache.put(8, "eight");
		memoryCache.put(9, "nine");
		memoryCache.put(10, "ten");
		memoryCache.put(11, "eleven");
		memoryCache.put(12, "twelve");
		memoryCache.put(13, "thirteen");
		memoryCache.put(14, "fourteen");
		memoryCache.put(15, "fifteen");
		memoryCache.put(16, "sixteen");
		memoryCache.put(17, "seventeen");
		memoryCache.put(18, "eighteen");
		memoryCache.put(19, "nineteen");
		memoryCache.put(20, "twenty");
		memoryCache.put(30, "thirty");
		memoryCache.put(40, "forty");
		memoryCache.put(50, "fifty");
		memoryCache.put(60, "sixty");
		memoryCache.put(70, "seventy");
		memoryCache.put(80, "eighty");
		memoryCache.put(90, "ninety");
		memoryCache.put(1000, "one thousand");
	}

	private String numberAsLetters(final int n) {
		if (memoryCache.containsKey(n)) {
			return memoryCache.get(n);
		}

		String string = n + "";
		if (n < 100) {
			memoryCache.put(n, memoryCache.get(10 * Character.getNumericValue(string.charAt(0))) + '-'
					+ memoryCache.get(Character.getNumericValue(string.charAt(1))));
		}

		if ((n < 1000) && (n >= 100)) {
			if (n % 100 == 0) {
				memoryCache.put(n, memoryCache.get(Character.getNumericValue(string.charAt(0))) + " hundred");
			} else {
				memoryCache.put(n, memoryCache.get(Character.getNumericValue(string.charAt(0))) + " hundred and "
						+ memoryCache.get(Integer.parseInt(string.substring(1, 3))));
			}
		}

		return memoryCache.get(n);
	}

	private int lengthOfString(final String string) {
		return string.replace(" ", "").replace("-", "").length();
	}

	/**
	 * Problem solution entry point
	 * 
	 * @param args
	 */
	public static void main(String[] args) {
		new P017().run();
	}

	@Override
	public Number solve() {
		problemNumber = 17;
		int result = 0;
		final int n = 1000;
		for (int i = 1; i < (n + 1); i++) {
			final String data = numberAsLetters(i);
			result += lengthOfString(data);
		}

		return result;
	}

}

File: typescript-project-euler-100/src/p017.ts
import { Solution } from "./Solution";

export class P017 extends Solution {

    memoryCache: Map<number, string> = new Map<number, string>();

    constructor() {
        super();

        this.memoryCache.set(1, "one");
        this.memoryCache.set(2, "two");
        this.memoryCache.set(3, "three");
        this.memoryCache.set(4, "four");
        this.memoryCache.set(5, "five");
        this.memoryCache.set(6, "six");
        this.memoryCache.set(7, "seven");
        this.memoryCache.set(8, "eight");
        this.memoryCache.set(9, "nine");
        this.memoryCache.set(10, "ten");
        this.memoryCache.set(11, "eleven");
        this.memoryCache.set(12, "twelve");
        this.memoryCache.set(13, "thirteen");
        this.memoryCache.set(14, "fourteen");
        this.memoryCache.set(15, "fifteen");
        this.memoryCache.set(16, "sixteen");
        this.memoryCache.set(17, "seventeen");
        this.memoryCache.set(18, "eighteen");
        this.memoryCache.set(19, "nineteen");
        this.memoryCache.set(20, "twenty");
        this.memoryCache.set(30, "thirty");
        this.memoryCache.set(40, "forty");
        this.memoryCache.set(50, "fifty");
        this.memoryCache.set(60, "sixty");
        this.memoryCache.set(70, "seventy");
        this.memoryCache.set(80, "eighty");
        this.memoryCache.set(90, "ninety");
        this.memoryCache.set(1000, "one thousand");
    }

    lengthOfString(data: string): number {
        return data.replace(/ /g, "").replace("-", "").length;
    }

    numberAsLetters(n: number): string {
        if (this.memoryCache.has(n)) {
            return this.memoryCache.get(n) as string;
        }

        const data = n + "";
        if (n < 100) {
            this.memoryCache.set(n, this.memoryCache.get(10 * Number(data.charAt(0))) + '-'
                + (this.memoryCache.get(Number(data.charAt(1)))) as string);
        }

        if ((n < 1000) && (n >= 100)) {
            if (n % 100 == 0) {
                this.memoryCache.set(n, this.memoryCache.get(Number(data.charAt(0))) + " hundred");
            } else {
                this.memoryCache.set(n, this.memoryCache.get(Number(data.charAt(0))) + " hundred and "
                    + (this.numberAsLetters(Number(data.substring(1, 3)))));
            }
        }

        return this.memoryCache.get(n) as string;
    }

    solve() {
        this.problemNumber = 17;

        let result = 0;
        const n = 1000;
        for (let i = 1; i < (n + 1); i++) {
            const data = this.numberAsLetters(i);
            result += this.lengthOfString(data);
        }

        return result;
    }
}
